hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go

```diff
@@ -3723,41 +3727,55 @@ func (r *HostedClusterReconciler) validateAWSConfig(hc *hyperv1.HostedCluster) e
 		if servicePublishingStrategy != nil && servicePublishingStrategy.Type != hyperv1.Route && servicePublishingStrategy.Type != hyperv1.LoadBalancer {
 			errs = append(errs, fmt.Errorf("service type %v with publishing strategy %v is not supported, use Route", hyperv1.APIServer, servicePublishingStrategy.Type))
 		}
-
 	} else {
 		if !hyperutil.UseDedicatedDNSForKASByHC(hc) && servicePublishingStrategy.Type != hyperv1.LoadBalancer {
 			errs = append(errs, fmt.Errorf("service type %v with publishing strategy %v is not supported, use Route or Loadbalancer", hyperv1.APIServer, servicePublishingStrategy.Type))
 		}
 	}
 
+	if hc.Spec.Platform.AWS.EndpointAccess == hyperv1.Public && servicePublishingStrategy.Type != hyperv1.Route {
+		errs = append(errs, fmt.Errorf("service type %v with publishing strategy %v is not supported in combination with a Public endpoint access, use Route", hyperv1.APIServer, servicePublishingStrategy.Type))
+	}
+
 	return utilerrors.NewAggregate(errs)
 }
 
-// validatePublishingStrategyMapping validates that each published serviceType has a unique hostname.
+// validatePublishingStrategyMapping validates:
+// 1. Each published serviceType has a unique hostname.
+// 2. APIServer has hostname field set if using Route.
 func (r *HostedClusterReconciler) validatePublishingStrategyMapping(hc *hyperv1.HostedCluster) error {
 	hostnameServiceMap := make(map[string]string, len(hc.Spec.Services))
-	for _, svc := range hc.Spec.Services {
-		hostname := ""
-		if svc.Type == hyperv1.LoadBalancer && svc.LoadBalancer != nil {
-			hostname = svc.LoadBalancer.Hostname
-		}
-		if svc.Type == hyperv1.Route && svc.Route != nil {
-			hostname = svc.Route.Hostname
-		}
+	var errs []error
 
+	for _, svc := range hc.Spec.Services {
+		hostname := servicePublishingStrategyHostname(svc)
 		if hostname == "" {
```

Comment:
nit: combine this if with the one in the next line

Reply:
I still need to `continue` when `hostname == ""` and serviceType is not APIServer. I will need to add another check for that if I combine here.